### PR TITLE
Compute local rank grid with `lazy_grid`

### PR DESCRIPTION
**Pull request by albert-de-montserrat**

### CODE REVIEW

**Comment by boriskaus** [comment]
`lazy_grid` and friends are routines that are useful also outside `JustRelax`. Would it make sense to put that in a different package (e.g., `CompGrids`)?

**Comment by albert-de-montserrat** [comment]
It would makes sense. What's the state of `CompGrids`? is it usable or needs a good update/rework?

**Comment by boriskaus** [comment]
it essentially contains nothing, so adding `lazy_grid` here as well as the `CreateArrays` function we discussed on Discord would be a great start. 
It would allow us to make `GeophysicalModelGenerator` compatible with `ParallelStencil` examples with a rather minor changes.

**Comment by albert-de-montserrat** [comment]
Cool! Once I have everything fully figured out and tested in JR I will see how to move it out with a proper API. `CompGrids` can be a starting point.

**Comment by boriskaus** [comment]
Can `heatdiffusion_PT!` deal with nonlinearities caused by latent heat? 
If yes, it could fully replace the corresponding routine in `MagmaThermoKinematics` (with the advantage that `JustRelax` runs with MPI as well).

**Comment by albert-de-montserrat** [comment]
Not in the current state, but it should be just some extra lines of code. I
will add it as soon as I can. Anything else needed for MTK?

File: miniapps/benchmarks/stokes2D/shear_band/ShearBand2D_MPI.jl [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+end
+
+N      = 32
+n      = N 
+nx     = n  # if only 2 CPU/GPU are used nx = 67 - 2 with N =128
+ny     = n
+figdir = "ShearBands2D"
+igg  = if !(JustRelax.MPI.Initialized())
+    IGG(init_global_grid(nx, ny, 1; init_MPI = true)...)
```
Comment (aelligp) [inline miniapps/benchmarks/stokes2D/shear_band/ShearBand2D_MPI.jl:226]:
```suggestion
    IGG(init_global_grid(nx, ny, 1; init_MPI = true, select_device=false)...)
```

File: miniapps/benchmarks/stokes2D/VanKeken.jl/VanKeken.jl [hunk 2]
```diff
@@ -7,7 +7,7 @@ backend = "CUDA_Float64_2D" # options: "CUDA_Float64_2D" "Threads_Float64_2D"
 # set_backend(backend) # run this on the REPL to switch backend
 
 # setup ParallelStencil.jl environment
-@static if occursin("CUDA", JustPIC.backend) 
+@static if occursin("CUDA", JustPIC.backend)
     model  = PS_Setup(:CUDA, Float64, 2)
     environment!(model)
 else
```
Comment (aelligp) [inline miniapps/benchmarks/stokes2D/VanKeken.jl/VanKeken.jl:13]:
```suggestion
else
       model = PS_Setup(:cpu,  Float64, 2)
```

File: miniapps/benchmarks/stokes3D/shear_heating/Shearheating3D.jl [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+using JustRelax, JustRelax.DataIO, JustPIC
+import JustRelax.@cell
+using CUDA
+
+## NOTE: need to run one of the lines below if one wishes to switch from one backend to another
+# set_backend("Threads_Float64_3D")
+# set_backend("CUDA_Float64_3D")
+
+# setup ParallelStencil.jl environment
+model = PS_Setup(:CUDA, Float64, 3)
```
Comment (aelligp) [inline miniapps/benchmarks/stokes3D/shear_heating/Shearheating3D.jl:12]:
```suggestion
model = PS_Setup(:cpu, Float64, 3)
```

File: miniapps/benchmarks/stokes3D/shear_heating/Shearheating3D.jl [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+            # ------------------------------
+
+      end
+
+      return nothing
+  end
+
+figdir   = "3D_Benchmark_Duretz_etal_2014"
+save_vtk = false # set to true to generate VTK files for ParaView
+n        = 64
```
Comment (aelligp) [inline miniapps/benchmarks/stokes3D/shear_heating/Shearheating3D.jl:355]:
```suggestion
n        = 16
```

File: miniapps/convection/Particles2D/Layered_convection2D.jl [hunk 5]
```diff
@@ -9,11 +6,11 @@ import JustRelax.@cell
 # set_backend("CUDA_Float64_2D")
 
 # setup ParallelStencil.jl environment
-model = PS_Setup(:CUDA, Float64, 2)
+model = PS_Setup(:Threads, Float64, 2)
```
Comment (aelligp) [inline miniapps/convection/Particles2D/Layered_convection2D.jl:9]:
```suggestion
model = PS_Setup(:cpu, Float64, 2)
```

File: miniapps/convection/Particles3D/Layered_convection3D.jl [hunk 6]
```diff
@@ -342,21 +340,21 @@ function main3D(igg; ar=1, nx=16, ny=16, nz=16, figdir="figs3D", do_vtk =false)
                 )
             end
 
-            xz_slice = ny >>> 1
+            slice_j = Int(ny ÷ 2)
```
Comment (aelligp) [inline miniapps/convection/Particles3D/Layered_convection3D.jl:343]:
```suggestion
            slice_j = ny >>> 1
```

revert back to original notation

**Review by aelligp — CHANGES_REQUESTED** [review]
looks good, just some minor typos/suggestions

**Review by aelligp — APPROVED** [review]

**Pull request merged**